doc/man/jal_purge.8
.TH JAL_PURGE 8
.SH NAME
.B jal_purge
\-
.SM JALoP
database purge utility
.SH SYNOPSIS
.B jal_purge
[\fIOPTIONS\fR...] [UUID] ... [UUID]
.SH "DESCRIPTION"
This manual page describes the
.BR jal_purge
command.
.PP
The
.B jal_purge
utility removes records from a JAL database. Without any options, the
.B jal_purge
tool reports whether the the records in the list of UUIDs of the specified type that would be removed,
but does not purge any records.
Only records that have been sent to at least one JALoP Network Store are considered safe for removal.

Note that the tool does not stop removing records when it finds a record that was never sent to a JALoP Network Store,
it will skip the unsent record and continue.
For example,
consider a database that contains records with the sequence numbers 1, 2, 3, 4, and 5,
where records 1, 3, and 5 have been delivered to at least one JALoP Network Store.
A call to
.B jal_purge
with no arguments would report that it is prepared to delete records 1, 3, and 5,
leaving records 2 and 4 untouched.
.SH OPTIONS
.TP
\fB\-t\fR, \fB\-\-type=T\fR
Search within the specified type.
\fBT\fR may be the letter \fIj\fR (for journal records),
\fIa\fR (for audit records),
\fIl\fR (for log records).
.TP
\fB\-b\fR, \fB\-\-before=B\fR
Remove all records with a timestamp before or equal to \fBB\fR that have been sent to at least one JALoP Network Store.
The timestamp must be specified as an XML schema date, time or dateTime string.
The xmlschema-2 document (http://www.w3.org/TR/xmlschema-2/) describes these formats.
When the time does not include a timezone offset,
it is interpreted as local time.
.TP
\fB\-d\fR, \fB\-\-delete\fR
Delete the records.
The
.B jal_purge
tool does not remove records that the JALoP Network Store has not sent to at least one JALoP Network Store.
.TP
\fB\-f\fR, \fB\-\-force\fR
When \fB\-\-delete\fR (\fB\-d\fR) is given,
force the deletion of records even when the JALoP Network Store has not sent them to at least one JALoP Network Store.
When given without \fB\-\-delete\fR (\fB\-d\fR) this will report the records that would be deleted.
.TP
\fB\-c\fR, \fB\-\-compact\fR
Compact the databases associated to the JAL record type (j/a/l) passed via \fB\-\-type\fR (\fB\-t\fR) and return empty pages to the filesystem.
.TP
\fB\-e\fR, \fB\-\-batch-size\fR
LMDB only - Specify the the number of records to purge per transaction. The default is 10000 records if not specified. The minimum value is 1 and maximum is 10000.
.TP
\fB\-s\fR, \fB\-\-compact-path\fR
LMDB only - Specify path of where the temporary LMDB database is copied while compacting.
.TP
\fB\-a\fR, \fB\-\-skip-process-check\fR
This skips the JALoP process running check on compact.  WARNING!! Using this setting can result in DB corruption if compact is performed against a database that is in use.
.TP
\fB\-h\fR, \fB\-\-home=P\fR
Specify the root of the JALoP database, defaults to
.I /var/lib/jalop/db
.TP
\fB\-n\fR, \fB\-\-version\fR
Output the version information and exit.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Output the UUID for each deleted record.
.TP
\fB\-x\fR, \fB\-\-detail\fR
Report more information about the records \fB jal_purge \fR is reviewing for deletion.
This reports the action to be taken (Delete, Forced Delete, Keep), the inbound state (Confirmed or Unconfirmed),
the outbound state (Unsent, Sent, Synced), the local insertion timestamp, and the local nonce for each record.
.SH "SEE ALSO"
.BR jald (8),
.BR jal-local-store (5),
.BR jal_dump (8),
.BR jal_subscribe (8)




ReleaseNotes.txt
Summary of release 1.3.1.0-beta changes (Fri Oct 31 2025) -
    - Added the new "jaldb_inline_filter" component developed in the Rust programming language. This is an inline filter process that shall perform syntactic and symantic validation of each message from the 'JALD' publisher to the JALDB database. The goal is to protect the JALDB database from a potentially compromised 'JALD' publisher that interacts with external entities. With this new inline filter process in place, the 'JALD' publisher no longer has direct Read/Write access to the JALDB database. Pre-built RPMs with Rust dependencies (crates) for RHEL 7/8/9 are included in <JALoP>/3rd-Party/rust/ within the repository. Details can be found in the README and JALoP User Guide.

Summary of release 1.3.0.0 changes (Fri Jul 11 2025) -
    - Added LMDB database support to store and manage JAL records.
    - Removed Berkeley DB (BDB) support. Note that RHEL 10 will not support BDB either.
    - Updated jaldb_tool to support file-system based database (created by jal_subscribe C and jnl_test Java Subscribers).

Summary of release 1.3.0.0-beta changes (Fri Apr 25 2025) -
    - Added LMDB database support to store and manage JAL records.

Summary of release 1.2.0.0 changes (Fri Aug 30 2024)
    - Added RHEL 9 support.
    - Added SECCOMP filters to applications (jal-local-store, jald, jal_subscribe). Allowed system calls can be listed in each application's configuration file. By default, configuration files have SECCOMP filtering disabled.
    - Added Linux Capability code to jal-local-store to enforce one-way communication between JAL Producers and the local store process.
    - Updated jal-local-store code so it can be configured and started as a systemd service. The JAL socket, SECCOMP system calls etc. can be configured via systemd. Added example systemd scripts and configuration files.
    - Added support for configurable digest algorithms - sha256, sha384, sha512 - for digest-challenge handshake, system metadata, and application metadata.
    - Updated RPMs in 3rd-party/ directory.
    - Removed RPATH or RUNPATH from build scripts in accordance with RTBv5 requirement OSSR-41.
    - Fix lint warnings/issues found in SPEC files and scripts.
    - Updated/added example self-signed and CA-signed certs for TLS connection.
    - Added/updated scripts and READMEs to generate example self-signed and CA-signed certs.
    - Add DB_RECOVER option to jal-local-store CLI and configuration file.
    - Bug fix: allow digest calculation for non-XML audit payload.
    - Fixed memory leaks, cleaned up jaldb_context.
    - Updated user guide, man pages, configuration files.

Summary of release 1.1.0.2 changes (Thu Nov 10 2022)
    - Clean up db_context upon channel close.
    - Cache eventList.xsd JAF schema for repeated use.
    - Minor python2/3 compatibility fix for doxygen scripts

Summary of release 1.1.0.1 changes (Thu Oct 13 2022)
    - Set DB_THREAD flag all the time to prevent db corruption error (#706).

Summary of release 1.1.0.0 changes (Fri Oct 07 2022)
    - Added RHEL 8 support.
    - Fixed issues in multiple subscribers running simultaneously.
        -- Fixed transferring duplicate records to live subscribers.
        -- Programmatically allocate a larger Vortex threadpool.
    - Improved jald daemonization.
        -- Use daemon() to redirect stdin/out/err to /dev/null, set sid, change directory to "/" instead of using separate instructions for these.
        -- Use dup2 to re-redirect stdout/err if log_dir has been configured.
        -- Create a PID file.
    - Converted scons build scripts from Python2 to Python3.
    - Upgraded JAF schema (eventList.xsd) to the latest version 1.0.5.
    - Fixed compiler warnings on RHEL 7 and RHEL 8.
    - Fixed unit tests.

Summary of release 1.0.0.5 changes (Mon Oct 11 2021)

    Upgraded from TLSv1 to TLSv1.2
    Added JALoPv1 TLS instructions.
    Updated python and scons version information.

Summary of release 1.0.0.4 changes (Sat May 01 2021)

    Refs #659: Fixed compiler warnings, and couple of broken unit tests caused by #ifdef SO_PEERCRED.

Summary of release 1.0.0.3 changes (Thu Dec 31 2020)

    Fixes use before initialization of jalp_context in jalp_test.
    Remove journal files from the disk that are purged via jal_purge.
    Fixes deadlock in the local_store and producer_lib involving multiple threads interacting with the libxm
l2 library.
    Fixed jal-local-store not shutting down upon SIGTERM in non-debug mode.
    Added flow-control configuration items.
    Shutdown jal-local-store upon issues with database.
    Added flow control to the jal-local-store by introducing accept delay for the UNIX Domain Socket. The accept delay is dynamically computed based on the current number of socket connections from the producers.
    Do not delete jal.socket created by a previous instance.
    Shutdown jal-local-store upon issues with database.
    Check if db_root exist and is a directory or return error.
    Added jaldb_iterate_by_timestamp2(*) function that creates a list of records to be purged and then remove them instead of re-initiating the database query after each record purge.
    Move setup_signal() call to the beginning of main(). Fix jal_purge not exiting gracefully on SIGTERM. 
    Create a list of records to be purged and then remove them instead of re-initiating the database query.

src/utils/jal_purge/jal_purge.cpp
/**
 * @file
 *
 * @brief This file contains the source for jal_purge
 *
 * ### LICENSE
 *
 * Source code in 3rd-party is licensed and owned by their respective
 * copyright holders.
 *
 * All other source code is copyright Tresys Technology and licensed as below.
 *
 * Copyright (c) 2012-2014 Tresys Technology LLC, Columbia, Maryland, USA
 *
 * This software was developed by Tresys Technology LLC
 * with U.S. Government sponsorship.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

#include <argp.h>
#include <iostream>
#include <algorithm>
#include <boost/filesystem.hpp>
#include <boost/algorithm/string/join.hpp>
#include <fstream>
#include <map>
#include <jalop/jal_version.h>
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include "jaldb_context.hpp"
#include "jaldb_purge.hpp"
#include "jaldb_record_dbs.h"
#include "jaldb_utils.h"
#include "jaldb_status.h"
#include "jaldb_strings.h"
#include "jaldb_record.h"
#include "jaldb_segment.h"
#include "jaldb_config.h"

#include "jal_alloc.h"
#include "jal_asprintf_internal.h"

using namespace std;

enum purge_action {
	JAL_PURGE_KEEP,
	JAL_PURGE_DELETE,
	JAL_PURGE_FORCE,
};

const char *send_str[] = { "UNSENT", " SENT ", "SYNCED" };
const char *recv_str[] = { "UNCONF", " CONF " };
const char *action_str[] = {"Keep  ", "Delete", "Force "};
static int exiting = 0;

const int MAX_PURGE_BATCH_SIZE = 10000;

//These are the jal processes to check if running before doing a db compact operation
const std::vector<std::string> jal_process_list = {"jal-local-store", "jald", "jal_subscribe", "jal_sub_cpp"};

static struct global_args_t {
	int del;
	int force;
	int verbose;
	int detail;
	int skip_clean;
	int compact;
	int skip_process_check;
	char *compact_path;
	char *batch_size;
	char *nonce;
	list<string> uuids;
	char type;
	char *before;
	char *home;
} global_args;

static void global_args_free();

// argp
const char *argp_program_version = jal_version_as_string();
const char *argp_program_bug_address = 0;
static char args_doc[] = "[uuid(s)]";
static char doc[] =
	"jal_purge - JALoP database purge utility.";
static error_t parse_opt(int key, char *arg, struct argp_state *state);
static struct argp_option options[] = {
	{"type", 't', "T", 0,
		"Specify the type of JAL record.  'T' may be 'j', 'a', or 'l' for journal, audit, or logging, respectively. When used with the '-u' or '--uuid' options, checks the timestamp, rather than the sequence ID, for record removal.", 0},
	{"before", 'b', "B", 0,
		"Remove all records with a timestamp before or equal to B. The timestamp must be specified as an XML schema date, time, or dateTime string. The xmlschema-2 document describes these formats. Only valid if no uuids are specified.", 0},
	{"delete", 'd', NULL, 0,
		"Delete the records. The jal_purge tool does not remove records that the JALoP Network Store has not sent to at least one JALoP Network Store.", 0},
	{"force", 'f', NULL, 0,
		"When '-d' is given, force the deletion of records even when the JALoP Network Store has not sent them to at least one JALoP Network Store.  When given without '-d', this will report the records that would be deleted.", 0},
	{"compact", 'c', NULL, 0,
		"Compact the databases associated to the JAL record type (j/a/l) passed via -t and return empty pages to the filesystem.", 0},
	{"batch-size", 'e', "E", 0,
			"Specify the the number of records to purge per transaction. The default is 10000 records if not specified. The minimum value is 1 and maximum is 10000.", 0},
	{"compact-path", 's', "S", 0,
			"Specify path of where the temporary LMDB database is copied while compacting.", 0},
	{"skip-process-check", 'a', NULL, 0,
			"This skips the JALoP process running check on compact.  WARNING!! Using this setting can result in DB corruption if compact is performed against a database that is in use.", 0},
	{"home", 'h', "H", 0, "Specify the root of the JALoP database, defaults to /var/lib/jalop/db.", 0},
	{"verbose", 'v', NULL, 0, "Output the UUID for each deleted record.", 0},
	{"detail", 'x', NULL, 0, "Report detailed information about the records jal_purge is reviewing for deletion. This reports the action to be taken (Delete, Forced Delete, Keep), the inbound state (Confirmed or Unconfirmed), the outbound state (Unsent, Sent, Synced), the local insertion timestamp, and the local nonce for each record.", 0},
	{NULL, 0, NULL, 0, NULL, 0}
};
static struct argp argp = {options, parse_opt, args_doc, doc, NULL, NULL, NULL};

static int setup_signals();
static void sig_handler(int sig);

extern "C" enum jaldb_iter_status iter_cb(const char *nonce, struct jaldb_record *rec, void *up);

bool isNotNumber(char c)
{
    return !(isdigit(c));
}

bool isNumber(const std::string& currDirName)
{
    return find_if(currDirName.begin(), currDirName.end(), isNotNumber) == currDirName.end();
}

bool checkForRunningProcess(const std::vector<std::string> &procNames, bool &isRunning)
{
	isRunning = false;

	std::string dirPath = "/proc";
	try
	{
		if (!boost::filesystem::exists(dirPath) || !boost::filesystem::is_directory(dirPath)) {
			fprintf(stderr, "ERROR: Failed to check for running JALoP process.\n");
			return false;
		}

		//Loop through all process id dirs in /proc and check if process listed in procName vector is running
		boost::filesystem::directory_iterator end_itr;
		for (boost::filesystem::directory_iterator dir_itr(dirPath); dir_itr != end_itr; ++dir_itr)
		{
			//Only process number directories
			std::string currDirName = dir_itr->path().filename().string();

			if (boost::filesystem::is_directory(dir_itr->path()) && isNumber(currDirName))
			{
				//Read contents of comm file in process subdir and check if it is one of the process names
				//provided in the procNames vector
				std::string currCommFile = dirPath + "/" + currDirName + "/comm";
				std::ifstream commFile(currCommFile);

				if (commFile.is_open())
				{
					std::string currCommand;
					while(std::getline(commFile, currCommand))
					{
						if (std::find(procNames.begin(), procNames.end(), currCommand) != procNames.end())
						{
							isRunning = true;
							break;
						}
					}

					commFile.close();

					if (true == isRunning)
					{
						break;
					}
				}
			}
		}
	}
	catch (std::exception &err)
	{
		fprintf(stderr, "ERROR: Failed to check for running JALoP process: %s\n", err.what());
		return false;
	}
	catch (...)
	{
		fprintf(stderr, "ERROR: Failed to check for running JALoP process - unknown error occurred\n");
		return false;
	}

	return true;
}

int main(int argc, char **argv)
{
	enum jaldb_status dbret = (enum jaldb_status)-1;
	enum jaldb_rec_type type = JALDB_RTYPE_UNKNOWN;
	jaldb_context *ctx = NULL;
	int err = 0;
	enum jaldb_flags jdb_flags = JDB_LMDB_PERFORMANCE_LEVEL2;
	std::string database_option = std::string(JDB_LMDB_PERFORMANCE_LEVEL2_STR);
	int map_size = DEFAULT_LMDB_MAP_SIZE;
	jaldb_config *jdb_config = NULL;
	enum jaldb_config_status jcs = JALDB_CONFIG_OK;

	// Perform signal hookups
	if ( 0 != setup_signals()) {
		goto out;
	}

	err = argp_parse(&argp, argc, argv, 0, 0, NULL);
	if(0 != err) {
		fprintf(stderr, "ERROR: Cannot parse command line arguments.\n");
		goto out;
	}
	if ((!global_args.uuids.empty() || global_args.before) && !global_args.type) {
		fprintf(stderr, "ERROR: -t required if -b specified.\n");
		goto out;
	}

	ctx = jaldb_context_create();
	if (!ctx) {
		fprintf(stderr, "Failed to create jaldb context\n");
		goto out;
	}

	if (global_args.compact) {
		//Ensure compact path was provided
		if (NULL == global_args.compact_path)
		{
			fprintf(stdout, "Compact path (-s) is required when compacting the database (-c)\n");
			global_args_free();
			jaldb_context_destroy(&ctx);
			return -1;
		}

		ctx->db_root = new std::string(global_args.home);
		ctx->compact_path = new std::string(global_args.compact_path);
	}
	else
	{
		//If not compact, disallow compact only args
		if (NULL != global_args.compact_path)
		{
			fprintf(stdout, "ERROR: Compact path (-s) is only valid when when compacting the database (-c)\n");
			global_args_free();
			jaldb_context_destroy(&ctx);
			return -1;
		}
	}

	//Sets batch_size if specified, otherwise default to no batching (1)
	ctx->batch_size = 1;

	if (NULL != global_args.batch_size)
	{
		std::string batch_size_str(global_args.batch_size);
		long curr_batch_size = 1;
		size_t pos;
		try
		{
			curr_batch_size = std::stol(batch_size_str, &pos);
		}
		catch(const std::invalid_argument& ia)
		{
			fprintf(stderr, "ERROR: Invalid batch_size argument.\n");
			goto out;
		}
		catch(const std::out_of_range& oor)
		{
			fprintf(stderr, "ERROR: batch_size argument out of range.\n");
			goto out;
		}

		if (curr_batch_size < 1)
		{
			fprintf(stderr, "ERROR: batch_size argument must be not be less than 1.\n");
			goto out;
		}

		ctx->batch_size = curr_batch_size;
	}

	if (global_args.compact) {
		//Ensure that jalop processes are not running (if skip process check not true)
		if (1 != global_args.skip_process_check)
		{
			bool isRunning = false;
			if (!checkForRunningProcess(jal_process_list, isRunning))
			{
				//Error occurred checking for process name, and error was printed
				//in method above.
				global_args_free();
				jaldb_context_destroy(&ctx);
				return -1;
			}

			if (true == isRunning)
			{
				std::string jal_process_list_str =  boost::algorithm::join(jal_process_list, ",");
				fprintf(stdout, "ERROR: compact cannot be performed due to one or more of the following JALoP processes are running: %s \n", jal_process_list_str.c_str());
				global_args_free();
				jaldb_context_destroy(&ctx);
				return -1;
			}
		}
	}
	else
	{
		if (1 == global_args.skip_process_check)
		{
			fprintf(stdout, "ERROR: Skip process check (-a) is only valid when when compacting the database (-c)\n");
			global_args_free();
			jaldb_context_destroy(&ctx);
			return -1;
		}
	}

	//Attempts to load optional LMDB_CONFIG file in db_root
	//If present, this will override the lmdb performance level
	//and lmdb map size, otherwise default values will be used.
	jcs = get_jaldb_config(global_args.home, &jdb_config);
	if (jcs != JALDB_CONFIG_OK && jcs != JALDB_CONFIG_E_NOTFOUND) {
		global_args_free();
		jaldb_context_destroy(&ctx);
		return -1;
	}

	//Only override map size if present in config
	if (jcs != JALDB_CONFIG_E_NOTFOUND)
	{
		if (jdb_config->map_size != 0)
		{
			map_size = jdb_config->map_size;
		}

		//Only override database option if present in config
		if (NULL != jdb_config->database_option)
		{
			jdb_flags = jdb_config->jdb_flags;
			database_option = std::string(jdb_config->database_option);
		}
		free_jaldb_config(&jdb_config);
	}

	dbret = jaldb_context_init(ctx, global_args.home, jdb_flags, map_size);
	if (JALDB_OK != dbret) {
		fprintf(stderr, "Failed to initialize jaldb context\n");
		goto out;
	}

	switch (global_args.type) {
	case 'j':
		type = JALDB_RTYPE_JOURNAL;
		break;
	case 'a':
		type = JALDB_RTYPE_AUDIT;
		break;
	case 'l':
		type = JALDB_RTYPE_LOG;
		break;
	default:
		goto out;
	}

	if (global_args.detail) {
		// Output the new detailed format
		printf("\nJAL_PURGE\n");
		printf("============\n");
		printf("SETTINGS:\n");

		if (type == JALDB_RTYPE_JOURNAL ) {
			printf("Journal records\n");
		} else if (type == JALDB_RTYPE_AUDIT) {
			printf("Audit records\n");
		} else if (type == JALDB_RTYPE_LOG) {
			printf("Log records\n");
		}

		if (global_args.del) {
			printf("Delete\n");
		} else {
			printf("Preview (no deletion)\n");
		}

		if (global_args.force) {
			printf("Synced + unsynced records\n");
		} else {
			printf("Synced records only\n");
		}

		if (global_args.batch_size)
		{
			printf("Batch Size: %s\n", global_args.batch_size);
		}
		else
		{
			printf("Batch Size: %d\n", MAX_PURGE_BATCH_SIZE);
		}

		printf("Database Option: %s\n", database_option.c_str());
		printf("LMDB Map Size (GB): %d\n", map_size);

	} else {
		// Otherwise output the old format that works with the test harness
		if (global_args.del) {
			cout << "Processing the following candidate records:" << endl;
		} else {
			cout << "Would process the following candidate records:" << endl;
		}
	}

	if (!global_args.uuids.empty()) {
		struct jaldb_record *rec = NULL;
		uuid_t uuid;
		char *nonce = NULL;
		list<string>::iterator iter;
		for(iter = global_args.uuids.begin(); iter != global_args.uuids.end(); iter++) {
#ifdef sun
			char * uuid_string = jal_strdup(iter->c_str());
			if (0 != uuid_parse(uuid_string, uuid)) {
				free(uuid_string);
#else
			if (0 != uuid_parse(iter->c_str(), uuid)) {
#endif
				fprintf(stderr, "ERROR: Invalid UUID: %s \n", iter->c_str() );
				goto out;
			}
#ifdef sun
			free(uuid_string);
#endif
			dbret = jaldb_get_record_by_uuid(ctx, type, uuid, &nonce, &rec);
			if (dbret != 0) {
				fprintf(stderr,"Cannot get record for UUID: %s\n", iter->c_str());
				// Treat UUID not found as a non-error condition
				dbret = JALDB_OK;
			} else {
				/* Inbound: records should be confirmed. Outbound: records should be synced. */
				/* Force flag causes sync flag to be ignored. */

				enum purge_action record_action = JAL_PURGE_KEEP;

				if (rec->confirmed && (rec->synced == JALDB_SYNCED)) {
					record_action = JAL_PURGE_DELETE;
				} else if (rec->confirmed && (rec->synced != JALDB_SYNCED) && global_args.force) {
					record_action = JAL_PURGE_FORCE;
				} else {
					record_action = JAL_PURGE_KEEP;
				}

				// If the detail flag is set, output the new detailed format, otherwise use the old format to prevent test harness from breaking
				if (global_args.detail) {
					// Print status of all records whether to be deleted or not
					if (global_args.del) {
						printf("%s %s %s %26s %s\n", action_str[record_action], recv_str[int(rec->confirmed)], send_str[int(rec->synced)], rec->timestamp, nonce);
					}
					else {
						printf("Preview: %s %s %s %26s %s\n", action_str[record_action], recv_str[int(rec->confirmed)], send_str[int(rec->synced)], rec->timestamp, nonce);
					}
				} else {
					printf("UUID: %s ", iter->c_str());
				}

				// Remove the record.
				if (global_args.del && rec->confirmed && (global_args.force || rec->synced == JALDB_SYNCED)) {
					dbret = jaldb_remove_record(ctx, type, nonce);
					if (dbret != 0) {
						fprintf(stderr, "ERROR: Cannot remove record: %s\n", nonce);
					} else {
						printf("Deleted\n");
					}
				} else {
					printf("Kept\n");
				}

				jaldb_destroy_record(&rec);
				free(nonce);
				nonce = NULL;
			}

			if (exiting) {
				break;
			}
		}
	} else if (global_args.before) {
		if (global_args.detail) {
			printf("Records before: %s\n\n", global_args.before);
		}
		dbret = jaldb_iterate_by_timestamp_purge(ctx, type, global_args.before, iter_cb, &global_args, exiting);
		goto out;
	} else {
		fprintf(stderr, "ERROR: Purging without a before time or uuid specified is currently not supported.\n");
		dbret = (enum jaldb_status)-1;
		goto out;
		// TODO: No arguments/filter specified
	}

out:
	if (global_args.detail) {
		printf("\n");
	}

	jaldb_status rc = JALDB_OK;
	if (global_args.compact) {
		fprintf(stdout, "Running DB->compact\n");
		rc = jaldb_compact_dbs(ctx, type);
		if (JALDB_OK != rc) {
			fprintf(stderr, "ERROR: Compact failed on one or more databases.\n");
		}
	}
	free((char *)argp_program_version);
	global_args_free();
	jaldb_context_destroy(&ctx);

	if (JALDB_OK != dbret || JALDB_OK != rc) {
		return -1;
	}
}

extern "C" enum jaldb_iter_status iter_cb(const char *nonce, struct jaldb_record *rec, void *)
{

	/* Inbound: records should be confirmed. Outbound: records should be synced. */
	/* Force flag causes sync flag to be ignored. */

	enum purge_action record_action = JAL_PURGE_KEEP;
	enum jaldb_iter_status ret_val = JALDB_ITER_CONT;
	char action[8] = {0};

	if (rec->confirmed && (rec->synced == JALDB_SYNCED)) {
		record_action = JAL_PURGE_DELETE;
		ret_val = JALDB_ITER_REM;
		strcpy(action, "Deleted");
	} else if (rec->confirmed && (rec->synced != JALDB_SYNCED) && global_args.force) {
		record_action = JAL_PURGE_FORCE;
		ret_val = JALDB_ITER_REM;
		strcpy(action, "Deleted");
	} else {
		record_action = JAL_PURGE_KEEP;
		ret_val = JALDB_ITER_CONT;
		strcpy(action, "Kept");
	}
	// If the detail flag is set, output the new detailed format, otherwise use the old format to prevent test harness from breaking
	if (global_args.detail) {
		// Print status of all records whether to be deleted or not
		if (global_args.del) {
			printf("%s %s %s %26s %s\n", action_str[record_action], recv_str[int(rec->confirmed)], send_str[int(rec->synced)], rec->timestamp, nonce);
		}
		else {
			printf("Preview: %s %s %s %26s %s\n", action_str[record_action], recv_str[int(rec->confirmed)], send_str[int(rec->synced)], rec->timestamp, nonce);
		}
	} else if (0 != strcmp(action, "Deleted")) {
		printf("NONCE: %s %s\n", nonce, action);

		if (global_args.verbose) {
			char uuid[37]; // UUID are always 36 characters + the NULL terminator
			uuid_unparse(rec->uuid, uuid);
			printf("UUID: %s %s\n", uuid, action);
		}
	}


	// If delete flag was not set, force iterator to return continue
	if (!global_args.del) {
		ret_val = JALDB_ITER_CONT;
	}

	return ret_val;
}

static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
	switch (key)
	{
		case 't':
			if ('j' != *arg && 'a' != *arg && 'l' != *arg) {
				fprintf(stderr, "Invalid type\n");
				goto err_out;
			}
			global_args.type = *arg;
			break;
		case 'b':
			global_args.before = strdup(arg);
			break;
		case 'd':
			global_args.del = 1;
			break;
		case 'f':
			global_args.force = 1;
			break;
		case 'h':
			global_args.home = strdup(arg);
			break;
		case 's':
			global_args.compact_path = strdup(arg);
			break;
		case 'e':
			global_args.batch_size = strdup(arg);
			break;
		case 'a':
			global_args.skip_process_check = 1;
			break;
		case 'v':
			global_args.verbose = 1;
			break;
		case 'x':
			global_args.detail = 1;
			break;
		case 'c':
			global_args.compact = 1;
			break;
		case ARGP_KEY_ARG:
			global_args.uuids.push_front(string(arg));
			break;
		default:
			return ARGP_ERR_UNKNOWN;
	}
	return 0;

err_out:
	argp_usage(state);
	exit(-1);
}

static void global_args_free()
{
	free(global_args.before);
	free(global_args.home);
	free(global_args.compact_path);
	free(global_args.batch_size);
}

static int setup_signals()
{
        struct sigaction action_on_sig;
        action_on_sig.sa_handler = &sig_handler;
        sigemptyset(&action_on_sig.sa_mask);
        action_on_sig.sa_flags = 0;

        if (0 != sigaction(SIGABRT, &action_on_sig, NULL)) {
                fprintf(stderr, "failed to register SIGABRT.\n");
                goto err_out;
        }
        if (0 != sigaction(SIGTERM, &action_on_sig, NULL)) {
                fprintf(stderr, "failed to register SIGTERM.\n");
                goto err_out;
        }
        if (0 != sigaction(SIGINT, &action_on_sig, NULL)) {
                fprintf(stderr, "failed to register SIGINT.\n");
                goto err_out;
        }
        return 0;

err_out:
        return -1;
}

static void sig_handler(__attribute__((unused)) int sig)
{
	fprintf(stdout, "\nCaught SIGTERM - exiting...\n");
	exiting = 1;
}
